### ISSUE TITLE
[$2000] Chat - Whitespace is not added after an emoji while inserting via native keyboard

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23658]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat
2. Insert a couple of emojis via emoji picker available in composer and send the message.
3. Now add a bunch of emojis via the emoji picker in native keyboard and send the message

## Expected Result:
 As per the issue https://github.com/Expensify/App/issues/11100 and slack discussion https://expensify.slack.com/archives/C01GTK53T8Q/p1662746279291419 a white space is expected after each emoji.

## Actual Result:
Whitespace is added only if emoji is added via emoji picker in composer and not while added using native keyboard.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.44-2
**Reproducible in staging?:** y 
**Reproducible in production?:** y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

[emoji.webm](https://github.com/Expensify/App/assets/93399543/f91ac459-e64c-49ba-8800-c31e129544c8)

https://github.com/Expensify/App/assets/93399543/b833037d-c850-44d0-805c-e8d6fd5e78fd

**Expensify/Expensify Issue URL:**
**Issue reported by:** @aswin-s
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690359926207719

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0188d7c4a5e13254b6</li>
        <li>Upwork Job ID: 1684628795620294656</li>
        <li>Last Price Increase: 2023-10-20</li>
<li>Automatic offers: </li>
<ul>
<li>aswin-s | Contributor | 27373872</li>
<li>aswin-s | Reporter | 27373874</li>
</ul></li>
    </ul>
</details>

**Comment by daordonez11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Experience of inserting emoji in chat is not unified!
### What is the root cause of that problem?
So I will open a new discussion in slack. @mallenexpensify proposed the space but this is only true when emoji is written with :emoji, when the emoji is picked from the picker menu in slack it is written without trail space. We would have to write custom code to "process" the written emoji and set a space in the `onChange` of the composer to unify but my proposal is to set the experience same as slack!
### What changes do you think we should make in order to solve the problem?
I would remove the trail from `onEmojiSelected` in `ReportActionCompose` and `ReportActionItemMessageEdit` since the implementation of :emoji already includes the space to unify the experience with slack and leave the EmojiPickerMenu to behave as the native one. Same as slack.

Video of solution working same as slack with native menu and EmojiPickerMenu:

https://github.com/Expensify/App/assets/13544910/cc484e2f-2b73-497f-ab50-d6192d2a33d0


### What alternative solutions did you explore? (Optional)
Update the `onChange` handler of `Composer` component, use a regex to validate if there is an emoji and include the space after it is included via native picker. Still I think altering native behavior wouldn't be expected but if it is desired we can implement the custom code!

**Comment by mollfpr** [comment]
@daordonez11 I still prefer consistency in our emoji addition, and having a space is what we have now. So we should fix it by adding an emoji from the native keyboard.

>To detect the insertion of an emoji, we must rely on onChange. Here, we should check whether nativeEvent.data is an emoji.

@samh-nl What's your plan for this? Will we replace the current `this.updateComment` with the used `onChange` from the composer?

**Comment by daordonez11** [comment]
Oh yeah I agree @mollfpr it was discussed in this thread https://expensify.slack.com/archives/C01GTK53T8Q/p1690483337635379 definitely the other proposal is the way to go. I'll check if there's a better proposal to be done.

**Comment by CortneyOfstad** [comment]
I'm heading OoO until 8/14, so reassigning BZ to keep an eye on this while I'm out 👍

**Comment by mollfpr** [comment]
Thanks @samh-nl. I'll try your updated proposal.

**Comment by ShogunFire** [comment]
Just in case, should we add a space when we paste emojis ? If it's the case we can simplify the proposal

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we add an emoji with the native keyboard no space is added

### What is the root cause of that problem?

There is no code to handle space when we add a native emoji

### What changes do you think we should make in order to solve the problem?

In updateComment of ReportActionCompose, we should make the diff between this.state.value and comment to see what has changed.
We can for example use the library diff. 
In package.json we can add it like this:     "diff": "^4.0.1",

Then to know what changed we can call diffChars like this:
`var diff = diffChars(this.state.value, comment);`
This will return an array of objects, to verify that the only thing that was added was an emoji, we can go through that array and verify that only one of the object was added and that this object value is an emoji.

Then we just need to add a space after that emoji.

The  code would look like this:

```
const diff = diffChars(this.state.value, comment);
var shouldAddSpaceToEmoji = false;
var commentWithSpaceAfterNativeEmoji = "";
diff.forEach((diffObject) => {
    if(diffObject.added){
        if(shouldAddSpaceToEmoji || !EmojiUtils.containsOnlyEmojis(diffObject.value);){
            shouldAddSpaceToEmoji = false;
            return;
        }
        
        shouldAddSpaceToEmoji = true;
        commentWithSpaceAfterNativeEmoji += (diffObject.value + " ");
    }
    else if(diffObject.removed){
        shouldAddSpaceToEmoji = false;
        return;
    }
    else {
        commentWithSpaceAfterNativeEmoji += diffObject.value;
    }
});

if(shouldAddSpaceToEmoji)
    comment = commentWithSpaceAfterNativeEmoji;
```

We can put this code in a new function addSpaceAfterNativeEmoji  for more clarity

### What alternative solutions did you explore? (Optional)

**Comment by ShogunFire** [comment]
@mollfpr I made a proposal without using onChange because I am wondering if using onChange will be regression free

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

After consideration, we shouldn't change the `onChangeText` to avoid regression.

@ShogunFire I'm not on board with adding a new package just for this case. I think we can check the emoji from the last char of the text from `updateComment`.

**Comment by ShogunFire** [comment]
@mollfpr I can do that quite easily but the problem with that method is that if there is an emoji at the end, we will add a space, now if the user wants to suppress the last part of the comment, let's say the emoji, he will suppress the space first which will trigger updateComment and add the space again, the user would be stuck. Would you like my proposal better if I didn't use any library (basically recreating a function to see what changed) ?

**Comment by mollfpr** [comment]
@ShogunFire I like to see how it goes. Let's put it in your alternative solution.

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23658#issuecomment-1675120445)

**Comment by ShogunFire** [comment]
This is very simplify version, I actually just made a diff of the emojis, we can check more things if necessary for example:

- Check that only one emoji was added
- Count the occurences of each emoji to know if one that existed before has more occurences now

Tell me if it's better

**Comment by CortneyOfstad** [comment]
@mollfpr any thoughts on the comments above from @samh-nl and @ShogunFire?

**Comment by mollfpr** [comment]
Sorry for the delay. I'll get the review done in the morning.

**Comment by CortneyOfstad** [comment]
@mollfpr just checking where we're at? TIA!

**Comment by mollfpr** [comment]
Sorry, I forgot to post the update.

@ShogunFire I'm curious why we need to loop over the diff char?

@samh-nl Is the problem you encounter can be solved? Or it's a regression from chancing the `onChangeText`?

**Comment by ShogunFire** [comment]
Hello @mollfpr Are you talking about the alternative proposal or the original one ? 

For the original one I think that was pretty clear

> This will return an array of objects, to verify that the only thing that was added was an emoji, we can go through that array and verify that only one of the object was added and that this object value is an emoji.

Have you also checked the alternative proposal without using libs ?

**Comment by ShogunFire** [comment]
Actually it's not that clear ahah, the array returned by diffChars looks like this.

String1 = "abc"
String2 = "abbc"

array returned by diffChars looks like this
["ab", "b" added, "c"]

**Comment by mollfpr** [comment]
@ShogunFire The alternative solution. I'm afraid a nested loop every time the user typing will affect the performance.

**Comment by zanyrenney** [comment]
Unassigning as per - https://github.com/Expensify/App/issues/23658#issuecomment-1677609191

**Comment by ShogunFire** [comment]
@mollfpr I understand but it only loops through the emojis so that should not be too much. I don't think there is a way to know the differences between two texts without nested loops. If we can make onChange works without regression I agree this is a better solution.

**Comment by ShogunFire** [comment]
If I resolve the regression can we share the bounty @samh-nl ? :P

**Comment by ShogunFire** [comment]
Is the isComposing boolean true for the second call also ? From what I have seen we can ignore the calls while it is composing.

If isComposing is true for the second call, we can calculate it ourself. I found this issue  in react native https://github.com/facebook/react/issues/3926

And there are a lot of solutions for example this one: 

https://github.com/facebook/react/issues/3926#issuecomment-1200414788

**Comment by ShogunFire** [comment]
My android is barely working now so it's hard to test sorry

**Comment by CortneyOfstad** [comment]
@mollfpr any updates here? TIA!

**Comment by mollfpr** [comment]
@CortneyOfstad We still don't have a satisfactory proposal, but the solution I am looking forward to is still under discussion by @samh-nl @ShogunFire 

Could we make this issue external again?

> InteractionManager.runAfterInteractions

@samh-nl Can we have the ready proposal now? Also, what is the delay purpose before adding the space?

**Comment by ShogunFire** [comment]
Then I think we can just ignore the call to onChange if isComposing is true, I am testing it

**Comment by ShogunFire** [comment]
I only have target, eventCount and text in nativeEvent

**Comment by ShogunFire** [comment]
@mollfpr Since onChange is not a possible solution (it looks like it doesn't have nativeEvent.data on some platforms) i think we have to make a diff of what changed so one of my solutions might still be a good one

**Comment by CortneyOfstad** [comment]
@mollfpr reapplied the `external` label 👍

**Comment by ShogunFire** [comment]
As I said since we can't use onChange, doing a diff is the only way to know that we added an emoji so my proposal is still valid, can we create a discussion on slack maybe to get some opinions ?

**Comment by mollfpr** [comment]
@ShogunFire Sure, you can start the discussion and send the link here.

I still do not understand why we need to get the diff; why can't we just check if the last text is an emoji or not?

**Comment by ShogunFire** [comment]
Because the user would not be able to delete this whitespace since it would be recreated when he does. 

Here is the l

**Comment by mollfpr** [comment]
>Because the user would not be able to delete this whitespace since it would be recreated when he does.

I can delete the whitespace on added emojis from our emoji picker. Is that different?

**Comment by mollfpr** [comment]
I'll review again the previous proposal from @ShogunFire. Meanwhile, we still open for new proposals.

**Comment by CortneyOfstad** [comment]
@mollfpr thoughts on the proposal above? TIA!

**Comment by mollfpr** [comment]
Sorry for the delay.

I need more time to decide on the @ShogunFire proposal.

For @tsa321 Could you explain what your regex does?

**Comment by ShogunFire** [comment]
@mollfpr Actually I thought of something else that is not perfect but makes everything simpler, we could test the last part of the newText and check if it's an emoji and if it is we can add a space. And to prevent adding a space when the user removes it we can check if the length of the new text is longer than the former text. I think we still need a regex because emojis are not stored in one char but that still more efficient and prettier than my other diff solutions

**Comment by CortneyOfstad** [comment]
@mollfpr ^^^

**Comment by mollfpr** [comment]
Sorry for the delay!

@tsa321 I still don't understand where the regex will be used; I don't see the use of ExpensiMark in the `ComposerWithSuggestions` component.

@ShogunFire That would be better; please update your proposal with it.

**Comment by mollfpr** [comment]
@tsa321 So the space will added on the comment sent? How will that solve this issue?

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23658#issuecomment-1675120445)

**Comment by mollfpr** [comment]
@ShogunFire Thanks for the update/

Number 2 is okay, but adding a space on emoji that it's already in the input seems a regression.

**Comment by ShogunFire** [comment]
@mollfpr Apart from doing a diff the only other way I see to avoid 1) would be to check the position of the cursor and only add the space if the cursor is at the end but the position of the cursor is updated in onSelectionChange and I think there is a bug where some platforms have onSelectionChange before onChangeText and some have it after :/

**Comment by ShogunFire** [comment]
Maybe the best way to do a diff is to save the emojis that we had in the previous text

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23658#issuecomment-1675120445)

Another try with a diff but this time I think it is efficient because we only do it if the number of emojis changed, also I believe we will need this diff for another bug that I found here: https://expensify.slack.com/archives/C049HHMV9SM/p1694526008852389

**Comment by mollfpr** [comment]
@ShogunFire Could you attach a video with the working solution?

**Comment by ShogunFire** [comment]
@mollfpr It doesn't work :( It's impossible to know which one we added if there are multiple times the same emoji because if we add an emoji in the middle it moves all the indexes

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23658#issuecomment-1675120445)

I hoped this is the last one, I included a video this time, it works fine, the only thing is it adds a space only if we are at the end of the string

**Comment by mollfpr** [comment]
@ShogunFire Does your solution only works with adding an emoji at the of the text?

**Comment by ShogunFire** [comment]
@mollfpr yes

**Comment by mollfpr** [comment]
@ShogunFire I thought the whitespace should be added on every position, not just for the emoji at the end of the text, right? Similar to what we have on the App emoji picker.


https://github.com/Expensify/App/assets/25520267/4aff5531-9154-47cb-804d-4cf7d30a8e2b

**Comment by ShogunFire** [comment]
Yes... I spent a lot of time on this but I don't manage to make that work :/

**Comment by mollfpr** [comment]
@CortneyOfstad Could we double this? No proposal seems right to me.

**Comment by mollfpr** [comment]
@tsa321 That's looks great!

>Is this result acceptable? Or should I fix the fast input emoji first?

I would prefer to have it working in any case.

**Comment by ShogunFire** [comment]
@tsa321 Also try it on mweb and safari there is an issue with onSelectionChanged being called before onChangeText on some platforms and that might break your solution

**Comment by CortneyOfstad** [comment]
Just checking to see where we're at here @tsa321 / @mollfpr? TIA!

**Comment by mollfpr** [comment]
> for mweb it is working fine, but not in ios msafari.

What's the issue on the safari? Does this relate to another issue that we are trying to fix in Safari?

>and I am still not able to fix the fast emoji input in android, I am not sure if I can fix it.

I don't feel this is okay; we should get it working on all platforms, except, we already handling a specific issue that eventually make this work.

@CortneyOfstad I think we can double this to get more eyes.

**Comment by CortneyOfstad** [comment]
Doubling it!

**Comment by mollfpr** [comment]
No desire proposals.

**Comment by ShogunFire** [comment]
@aswin-s I thought about doing this but it doesn't work when you add multiple times the same emoji.
For example if the text before was 🙂 and now it's 🙂🙂 you have no way of knowing if you added the first emoji or the second. With your method you will add the space after the  second one but maybe the user added the first one.

**Comment by CortneyOfstad** [comment]
@mollfpr thoughts on the proposals and updates above? TIA!

**Comment by mollfpr** [comment]
I'll get the proposal reviewed tomorrow. Thank you!

**Comment by CortneyOfstad** [comment]
Thanks @mollfpr!

**Comment by CortneyOfstad** [comment]
@mollfpr any updates on this? Thanks!

**Comment by CortneyOfstad** [comment]
Bump @mollfpr ^^^ thanks!

**Comment by mollfpr** [comment]
Sorry for the delay; I just running out of time. I will prioritize this tomorrow 🙏

**Comment by mollfpr** [comment]
@aswin-s Here's what I found on testing your proposal.

1. The cursor placement is not working well on selecting the emoji from the picker.

https://github.com/Expensify/App/assets/25520267/62c10f60-f9dc-4934-8c26-711e687ae56c

2. Select an emoji with the cursor selection not adding a space afterward, but the cursor placement is wrong. For the emoji native keyboard, it's only not adding space, but the cursor placement is correct.

**Comment by aswin-s** [comment]
@mollfpr There was duplicate code for fixing cursor position while entering emoji via EmojiPicker. I have [updated my proposal](https://github.com/Expensify/App/issues/23658#issuecomment-1744468391) to specify which lines to remove. Also there was a small bug while pasting over emoji over selected text. I've fixed both issues. Please retest. 

https://github.com/Expensify/App/assets/6880914/78fb1540-e013-4a99-8098-8584db922b24

**Comment by mollfpr** [comment]
@aswin-s Tested on all platforms, but the result on iOS is not working well. The space is only added on the first emoji and the emoji picker button and the back button is stuck afterward.

https://github.com/Expensify/App/assets/25520267/e1555013-72e1-4c22-9b70-e01392fe488a

**Comment by aswin-s** [comment]
@mollfpr Turns out this was caused due to the  usage of utility method `getCommonSuffixLength`. When we pass an empty string to this method, the while loop will run infinitely. This caused the native keyboard to hang. That's why you were seeing irresponsive keyboard and backspace was not working.

We'll need to fix `getCommonSuffixLength` method to return 0 if either of those strings are empty. 

https://github.com/Expensify/App/blob/7b03a4d23b0165dab450bfee270c5a75698aa82f/src/libs/ComposerUtils/index.ts#L33-L39

```jsx
function getCommonSuffixLength(str1: string, str2: string): number {
    let i = 0;

    // Add an early return if either of the strings are empty
    if(str1.length===0 || str2.length===0) {
        return 0;
    }

    while (str1[str1.length - 1 - i] === str2[str2.length - 1 - i]) {        
        i++;
    }
    return i;
}
```

**Comment by aswin-s** [comment]
[Updated my proposal](https://github.com/Expensify/App/issues/23658#issuecomment-1744468391) to include this patch as well. With the patch even iOS is behaving as expected.


https://github.com/Expensify/App/assets/6880914/7bf354cf-e0e6-445d-b4ec-47714789e8e1

**Comment by CortneyOfstad** [comment]
I'm heading OoO until Oct. 25, so reassigning BZ to keep an eye on things while I'm gone 👍

**Comment by mollfpr** [comment]
https://github.com/Expensify/App/assets/25520267/4bcfe2f6-0656-4b9d-8ffd-5fc37d666f77

@aswin-s I still see the same result with the iOS. Do you mind creating your branch so I can test it precisely?

**Comment by aswin-s** [comment]
@mollfpr Here is my branch https://github.com/aswin-s/App/tree/fix/naitve-emoji-picker

**Comment by mollfpr** [comment]
@aswin-s I tested your branch, and it seems to work fine on all platforms. However, after adding the emoji from the emoji picker on Android and iOS, I found a wrong selection set.


https://github.com/Expensify/App/assets/25520267/44263646-f60c-4c86-888b-287c44d42d70


https://github.com/Expensify/App/assets/25520267/949ce97c-7639-45b8-8f2c-c4f9cd475c5e

**Comment by mollfpr** [comment]
@aswin-s Okay, it's confirmed the selection issue is already there. I'll get my final thoughts tomorrow.

**Comment by CortneyOfstad** [comment]
@isabelastisser I am back from OoO so I can take this back over — thanks for keeping an eye on things!

@mollfpr Added the **engineering** label, so that is all set 👍

**Comment by CortneyOfstad** [comment]
@mollfpr @flodnv any updates on this? Just checking to see where we are at 👍 Thanks!

**Comment by barros001** [comment]
I'm working on an issue that's also related to emojis and I was made aware of this issue. While testing to make sure my changes did not break the changes implemented by #30412 I noticed a few potential problems:

* If you paste an emoji, such as 😂, it does add the space. But if you do paste any string ending in an emoji, such as `abc 😂`, it does't. We do have several places that add spaces after emojis, I believe we want it to be added here as well;
* If you paste an emoji and a space, such as `😆 `, it will add another space at the end, making it two spaces;
* If you paste the same emoji several times in a row,  quickly, sometimes 2 spaces are added to the end (after a small delay). Some other times a random double space is added in the middle of the sequence of emojis. This doesn't happen all the time but I definitely saw this happening a few times to me (Safari web).

I wanted to post a video but my local dev decided to stop working just now when I went to record it 😞

**Comment by barros001** [comment]
I was finally able to record a video:


https://github.com/Expensify/App/assets/765936/50daf829-d1e2-4f8a-a1a1-7fd1feec8a06

**Comment by aswin-s** [comment]
@barros001 Thanks for reporting this.

> If you paste an emoji, such as 😂, it does add the space. But if you do paste any string ending in an emoji, such as abc 😂, it does't. We do have several places that add spaces after emojis, I believe we want it to be added here as well;

No we don't want to add a white space to anything other than an explicit emoji character. This feature is to assist user when he inserts an emoji through native keyboard, picker or short code. Pasting anything else is not supported at all. That would be breaking user expectation.

>If you paste an emoji and a space, such as 😆 , it will add another space at the end, making it two spaces;
If you paste the same emoji several times in a row, quickly, sometimes 2 spaces are added to the end (after a small delay). Some other times a random double space is added in the middle of the sequence of emojis. This doesn't happen all the time but I definitely saw this happening a few times to me (Safari web).

This one is a bug. Looks like the Emoji utility function `EmojiUtils.containsOnlyEmojis` trims whitespace from the text before checking for emoji characters. So it returns true even for inputs with an emoji and a trailing or leading whitespace. That's the reason extra white space is getting added. But since the utility function is used in many other aspects I wouldn't want to change its behaviour. Instead I'll check if the pasted content has a leading or trailing whitespace. 

@mollfpr I'll raise a followup PR shortly.

**Comment by barros001** [comment]
> No we don't want to add a white space to anything other than an explicit emoji character. This feature is to assist user when he inserts an emoji through native keyboard, picker or short code. Pasting anything else is not supported at all. That would be breaking user expectation.

We have some inconsistent behavior around how this handled that needs to be addressed. On staging, try pasting  the following:

`hello :joy:`
`hello 😂`

First one will add space at the end, second won’t. We have code that was specifically added to add a space after the last emoji converted either when pasting or entering text, that’s why I assumed we always want to add a space after an emoji of this emoji is the last character either manually entered or pasted.

**Comment by aswin-s** [comment]
Not sure how to handle above scenario. Personally I wouldn't prefer adding whitespace to anything other than an emoji. The real use case of the feature is to assist the user while typing in the composer. For pasting text, expectation is to match the copied text. In that sense we shouldn't be adding whitespace even to short code if it is accompanied by other text. 

Requesting your thoughts @mollfpr @flodnv

**Comment by mollfpr** [comment]
>If you paste an emoji, such as 😂, it does add the space. But if you do paste any string ending in an emoji, such as abc 😂, it does't. We do have several places that add spaces after emojis, I believe we want it to be added here as well;

This seems inconsistent, but I think we can fix it in another issue. In this issue, we should be focused on adding the whitespace when the emoji is added from the native keyboard or emoji picker.

**Comment by aswin-s** [comment]
@mollfpr As this is marked as a deploy blocker I want to confirm the expectations here. Right now whitespace is added if the last character inserted into the composer is an emoji. It doesn't differentiate whether the character was inserted via keyboard, picker or directly from clipboard. So should we limit it just to keyboard events and picker?

**Comment by mollfpr** [comment]
> So should we limit it just to keyboard events and picker?

I would say yes. The solution is already complicated, and since it's changing our `Composer` component, I'd rather do this to prevent more blockers.

@aswin-s May I know how the effort for either make the whitespace is added in all cases, and limit the feature only adding emoji from keyboard or emoji picker?

**Comment by aswin-s** [comment]
@mollfpr There is no direct way to detect paste event from a TextInput component. Some approaches include comparing changed text with contents in clipboard. But this approach is bound to fail if user has the same emoji copied to clipboard and then inserting the same emoji via soft keyboard. Not sure if it is worth the added complexity.

**Comment by mollfpr** [comment]
Okay, we need more discussion on whether the whitespace should be added from pasted emoji. Still, this issue already has a clear expected result: adding the whitespace on adding emoji from the native keyboard and keeping the current behavior on the emoji picker and emoji suggestion.

---

I think we should revert the PR for now. It might take some time to find a solution for the blockers.

**Comment by bondydaa** [comment]
@mollfpr agreed, let's revert for now. do you want to create the revert PR please?

**Comment by mollfpr** [comment]
@bondydaa Sorry, I'll be unavailable in a few minutes it's almost midnight for me 🙏

**Comment by bondydaa** [comment]
revert pr is up https://github.com/Expensify/App/pull/31011

**Comment by barros001** [comment]
> Okay, we need more discussion on whether the whitespace should be added from pasted emoji. Still, this issue already has a clear expected result: adding the whitespace on adding emoji from the native keyboard and keeping the current behavior on the emoji picker and emoji suggestion.

I'm gonna chime in here since I've been working around composer, specifically on issues related to Emojis. Before anything, I want to say that I'm coming from a place of respect and have no intention of stepping on people's toes or to hijack this issue (which has already been assigned). I stumbled upon this by accident and I have some ideas that I believe can help the project.

My understanding of the requirements is as follows: add a space after an Emoji is entered into Composer, either coming from the native keyboard, short-code, pasting short-codes or pasting native Emojis, with or without extra text preceding it. Either this is really what we want or not I'm not sure, but that's how the system works today.

There are 3 places where we handle this:

https://github.com/Expensify/App/blob/8c8ba52ba58699fa038b74079886674f9ffa93f3/src/libs/EmojiUtils.ts#L356-L360

This adds a space after the last emoji converted from the input (either typing or pasting a text)

https://github.com/Expensify/App/blob/8c8ba52ba58699fa038b74079886674f9ffa93f3/src/pages/home/report/ReportActionItemMessageEdit.js#L338-L344

This forces a space after an Emoji added from the Emoji picker, when editing a message

https://github.com/Expensify/App/blob/8c8ba52ba58699fa038b74079886674f9ffa93f3/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L319-L330

And this will add a space as well after emoji selected from the emoji picker, when composing a new message

These are the 3 places I see this rule, if you know of more places, please let me know. To me the fact that we have the same rule replicated in several different places, trying to account for difference scenarios doesn't look right. 

The issue I'm working on (https://github.com/Expensify/App/issues/29795), which by the way a proposal hasn't yet been picked (and I have no clue if mine will even be the one) deals with an issue where the cursor gets positioned on the wrong spot after converting a emoji short-code pasted into composer when it has more text after it. I proposed to change how we calculate the position of the cursor and to get rid of the `ComposerUtils.getCommonSuffixLength`, which IMO if flawed and does not work under certain scenarios.

The idea is to look at the cursor position before the text update, the difference in number of characters between previous and new text and then move the cursor forward or back that number of characters (there's some extra rules about emoji short-code as well, but that's not the point here). That way we don't have to compare two strings and determine what has changed (a tasks that could actually be impossible), but rather we just move the cursor forward or back based on the difference in size of the strings.

Anyways, the reason I'm bringing this up into here (an issue that has nothing really todo with the issue I'm working on), is that I realized that we can potentially get rid of these scattered rules that add space after emojis and centralize it on the `updateComment` method, similarly to what was tried on the PR associated with this issue. But instead of comparing the original and current text, we look at where the cursor is after the text update. Here's how it would work, in simple terms:

* Is the cursor immediately after an Emoji?
* Is current text longer than previous text? (if you're adding chars, we can safely append a space); or
* Did we NOT delete a space or series of spaces before landing after an emoji? (this will allow you to hit backspace and delete a space without a space being added again)
* We could even check if the next character after current cursor position is not a space?

If these conditions are met, then we add a space and push the cursor one step forward.

Doing this on `updateComment` (and the corresponding method on Edit) allow us to remove these hardcoded rules to add spaces from other places around the codebase. As a bonus, it will append space under ALL scenarios:
* Emoji via native keyboard; check
* Pasting emoji; check
* Pasting emji short-code; check
* Doing anything else where you end up with an Emoji immediately before the cursor; check

All this implies we want to always append a space after an emoji that is added to composer regardless of the input method (as per a [comment](https://github.com/Expensify/App/issues/23658#issuecomment-1798686792) above, we have no way to tell whether it was typed or pasted, so we might not have an option here)

While working on my proposal for the other issue I took the liberty to create a POC (could use some polishing and testing) for the idea I described above, you can view it here: https://github.com/Expensify/App/compare/main...barros001:expensify-app:paste-emoji-issue?expand=1

Here's a couple of videos I made showing the results:

https://github.com/Expensify/App/assets/765936/1f5eda9d-57ae-4bac-92c1-35b36f424fdc

https://github.com/Expensify/App/assets/765936/3d124b58-1e27-4833-9733-2ef527c42682

Anyways, I thought the issue I was working on had some overlap with this one and my proposal would build some foundation that could be built upon to solve this issue. I figure it'd be a nice challane to try something and get to learn more about the codebase in the process. Hopefully this can be of any help.

**Comment by aswin-s** [comment]
> Okay, we need more discussion on whether the whitespace should be added from pasted emoji. Still, this issue already has a clear expected result: adding the whitespace on adding emoji from the native keyboard and keeping the current behavior on the emoji picker and emoji suggestion.
> 
> I think we should revert the PR for now. It might take some time to find a solution for the blockers.

@mollfpr I've PR https://github.com/Expensify/App/pull/30955 up with fixes for emoji's with skin tone and extra whitespaces getting added. For handling whitespace after 'pasted' emojis I guess we need more discussion. 

@barros001 Thanks for the inputs. The approach sounds similar to the one in my proposal. But yes there is a scope for unifying the scattered logic and handling everything in one place.

**Comment by flodnv** [comment]
Closing as won't fix.

**Comment by barros001** [comment]
I was waiting for [one of my PR](https://github.com/Expensify/App/pull/31279) to be merged to comment here and I just got notification this issue was closed. Not meaning to re-open it but I just wanted to share some work I did on the side trying to solve all issues with emojis and spaces. Here's a PR I made against my own fork:

https://github.com/barros001/expensify-app/pull/1/files

Basically this will centralized where we decide whether we want to add space after the emoji in a single place, based on input changes. Once anything on composer changes, it will extract exactly what changed (using the `calculateTextDiff` function) and if we're **adding** text that **ends in an emoji**, it adds a space. It will also account for when a space is already present after the emoji and not add it again, but still move the cursor one step ahead.

This accounts for all types of emojis: short-code, native, typed manually or pasted.

It performed well on my tests but it might need a little more work. I'm not entire sure what `convertToLTRForComposer` and why we have it there and it could have an impact. Also, it will also conflict with the proposed solution for [this issue](https://github.com/Expensify/App/issues/28563), re-creating it.

Anyways, I'm not suggesting we use this work, it was more of a personal challenge while learning the codebase, but I thought I'd share it here as I believe this could come handy eventually if we decide to tackle this issue in the future.

**Comment by flodnv** [comment]
Interesting, thanks for sharing your work!

**Comment by aswin-s** [comment]
+1 @mollfpr I would appreciate some clarification on why the issue was closed. I understand that there was a regression in my initial solution, but I quickly fixed it and requested feedback. It seems unfair to close it without any explanation. I believe there is a policy on how to handle issues that gets closed after assignment.

**Comment by flodnv** [comment]
I assume we are talking about https://github.com/Expensify/App/pull/30955? I missed that, sorry, are we waiting on @mollfpr to review that?

**Comment by aswin-s** [comment]
@flodnv Yes that's the one. @mollfpr Could you please have a look. Also we still need a discussion on [adding whitespace after pasted emoji](https://github.com/Expensify/App/issues/23658#issuecomment-1799029415). An [issue was raised](https://github.com/Expensify/App/issues/30983) claiming that's not the expected scenario. However the intention of this PR is to append a whitespace after an emoji no matter how it gets inserted. Also detecting paste event from within a text input component is not trivial and will make the logic unnecessarily complex.

**Comment by aswin-s** [comment]
@mollfpr Is there anything I can assist you with to help this move forward?

**Comment by CortneyOfstad** [comment]
@mollfpr bump ^^^ Thanks!

**Comment by CortneyOfstad** [comment]
@mollfpr bump again — TIA!

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

I'm starting to work on this issue again and testing the new PR.

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

New feature.


>[@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1. Open any chat
2. Enter an emoji via a native keyboard, emoji picker, or short code (eg- :heart:)
3. Verify that the emoji gets appended by a white space.
4. Verify that emojis can be removed by tapping backspace.
5. Verify that the cursor is positioned properly after inserting the emoji.
6. Verify that emoji can be inserted in the midst of text
7. Verify that an emoji can be copy pasted
8. Verify that complex emoji like '👨‍👩‍👧‍👦' can be added and removed.
9. Verify that while pasting text, white space is added only when emoji is pasted. Whitespace shouldn't be added if the pasted text has any other characters other than emoji, including whitespace.
10. Verify that emojis with different skin tones can be inserted

---

@CortneyOfstad Could you give the payment summary for manual request? Thank you!

**Comment by CortneyOfstad** [comment]
@aswin-s — I sent you both offers as the reporter and the contributor. Please let me know once you accept those and I'll get them paid ASAP. 

@mollfpr payment summary is below!

**Comment by CortneyOfstad** [comment]
## Payment Summary

@aswin-s (reporter) — to be paid $50 via Upwork
@aswin-s (contributor) — to be paid $2000 via Upwork
@mollfpr (C+) — to be paid $2000 via NewDot

**Comment by aswin-s** [comment]
> ## Payment Summary
> 
> 
> 
> @aswin-s (reporter) — to be paid $50 via Upwork
> 
> @aswin-s (contributor) — to be paid $2000 via Upwork
> 
> @mollfpr (C+) — to be paid $2000 via NewDot

@CortneyOfstad The issue was reported before the bounty for issues were reduced to 50$.

**Comment by CortneyOfstad** [comment]
Regression test listed [here](https://github.com/Expensify/Expensify/issues/367572) 👍

**Comment by CortneyOfstad** [comment]
Sorry about that @aswin-s! I adjusted the payment to be $250, and will update the payment summary above. You should have received both payments successfully, totaling $2,250 👍

**Comment by CortneyOfstad** [comment]
Updated payment summary below:

## Payment Summary

@aswin-s (reporter) — paid $250 via Upwork ✅
@aswin-s (contributor) — paid $2000 via Upwork ✅
@mollfpr (C+) — to be paid $2000 via NewDot

**Comment by situchan** [comment]
We had to [revert](https://github.com/Expensify/App/pull/31011) first [PR](https://github.com/Expensify/App/pull/30412) due to regressions: #31006 #30950 #30983
2nd [PR](https://github.com/Expensify/App/pull/30955) also caused [regression](https://github.com/Expensify/App/issues/35339) but we closed it as minor.

Hope this info won't affect payment but worth mentioning.

**Comment by CortneyOfstad** [comment]
Alright, @aswin-s I requested a refund of 50% ($1000) as the contributor. Please let me know as soon as you complete this. Thank you!

**Comment by aswin-s** [comment]
@CortneyOfstad Let me know if I can do anything from my end

**Comment by aswin-s** [comment]
> Alright, @aswin-s I requested a refund of 50% ($1000) as the contributor. Please let me know as soon as you complete this. Thank you!

Thank you

**Comment by CortneyOfstad** [comment]
Thank you so much @aswin-s — I greatly appreciate it!

**Comment by CortneyOfstad** [comment]
Alright — here is the FINAL payment summary 😂 

## Payment Summary

@aswin-s (reporter) — paid $250 via Upwork ✅
@aswin-s (contributor) — paid $1000 via Upwork ✅  (50% cut due to regression)
@mollfpr (C+) — to be paid $1000 via NewDot (50% cut due to regression)

**Comment by aswin-s** [comment]
@CortneyOfstad @situchan Quick question. The regression was caught during staging itself and fix was raised the very next day. So payment would be deducted even if developer fixes it even before it hits PROD? Just asking so that I can be more careful next time.

**Comment by CortneyOfstad** [comment]
@aswin-s Thanks for asking! Each error in proposed PRs will carry different weights based on the type of error and the impact it would have, so in some cases (like this one) yes, it will cause a reduction in payment.

**Comment by situchan** [comment]
There are some exceptions. i.e. regressions penalties won't apply if already discussed as out of scope in PR.
And too much efforts working on PR.
i.e. Ideal Nav PR caused so many regressions but it was super large and reviewers won't get payment deducted.
But usually penalty is applied as long as bug is caught after PR is merged, whether in main branch, staging, or production

**Comment by JmillsExpensify** [comment]
$1,000 approved for @mollfpr based on [this summary](https://github.com/Expensify/App/issues/23658#issuecomment-1932318251).
